### PR TITLE
Add post processing algorithm based on work by Hardt, Price, Srebro

### PR DESCRIPTION
**Pull request by romanlutz**
This includes preliminary tests, separate plotting utilities.

What this does not include: APIs as discussed on Friday.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 * registered the project at [PyPI](https://pypi.org/)
 
-* changed how the fairness constraints are initialized (in `fairlearn.moments`), and how they are passed to the fair learning reduction `fairlearn.classred.expgrad`
+* changed how the fairness constraints are initialized (in `fairlearn.expgrad.moments`), and how they are passed to the fair learning reduction `fairlearn.expgrad.classred.expgrad`
```
Comment (riedgar-ms) [inline CHANGES.md]: Should changes like this be a separate PR?

File: README.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ A Python package that implements the black-box approach to fair classification d
 
 The package can be installed via `pip install fairlearn`. To verify that it works, download `test_fairlearn_smoke.py` from the repository and run `python -m pytest test_fairlearn_smoke.py`.
 
-Instead of installing the package, you can clone the repository locally via `git clone git@github.com:Microsoft/fairlearn.git`. To verify that the package works, run `pip install -r requirements.txt` for setup and to test `python test_fairlearn.py` in the root of the repository.
+Instead of installing the package, you can clone the repository locally via `git clone git@github.com:Microsoft/fairlearn.git`. To verify that the package works, run `pip install -r requirements.txt` for setup and to test `python test/test_fairlearn.py` in the root of the repository.
```
Comment (riedgar-ms) [inline README.md]: Should just need `pytest test/` to run all the tests

File: fairlearn/post_processing/roc_curve_based_post_processing.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    data = pd.DataFrame({ATTRIBUTE_KEY: attributes, SCORE_KEY: scores, LABEL_KEY: labels})
+
+    return data.groupby(ATTRIBUTE_KEY)
+    
+
+def roc_curve_based_post_processing_demographic_parity(attributes, labels, scores, gridsize=1000, flip=True, debug=False):
+    """ TODO add description
+    :param attributes: the protected attributes
+    :type attributes: list
+    :param labels: the labels of the dataset
```
Comment (riedgar-ms) [inline fairlearn/post_processing/roc_curve_based_post_processing.py]: Don't the labels have to be 0 or 1 for some of the other methods to work? I don't think this is checked anywhere in the code.

**Comment by riedgar-ms** [comment]
I don't think any of my comments should be taken as blocking for checkin. I'm sure that there will be future reworking and test adding, but since we confirmed that the basic algorithm is working, our highest priority is probably agreeing our API.

File: fairlearn/post_processing/roc_curve_based_post_processing.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    # labels provides better accuracy.
+    i = 0
+    count = [0, 0]
+    while i < n:
+        threshold = scores[i]
+        while scores[i] == threshold:
+            count[labels[i]] += 1
+            i += 1
+        # For the ROC curve we calculate points (x, y), where x represents
+        # the conditional probability P[Y_hat=0 | Y=0] and y represents
```
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/roc_curve_based_post_processing.py]: yes, not entirely sure what I was thinking there

File: fairlearn/post_processing/roc_curve_based_post_processing.py (hunk 3 above)
Comment (romanlutz) [reply fairlearn/post_processing/roc_curve_based_post_processing.py]: it is now!

File: fairlearn/post_processing/_roc_curve_utilities.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+def _get_roc(data, x_grid, attribute, flip=True):
+    """Get ROC curve's convex hull based on data columns 'score' and 'label'
+    Scores represent output values from the model.
+    """
+    roc_sorted = _calculate_roc_points(data, attribute, flip)
+    roc_selected = _filter_points_to_get_convex_hull(roc_sorted)
+    roc_convex_hull = pd.DataFrame(roc_selected)[['x', 'y', 'operation']]
+    return roc_convex_hull
+
+def _filter_points_to_get_convex_hull(roc_sorted):
```
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/_roc_curve_utilities.py]: I'm adding two more tests so we have a variety of scenarios covered (0, 1, and multiple dropped points)

[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/_roc_curve_utilities.py]: actually it ended up being very simple to think of more cases so I just added them

File: fairlearn/post_processing/_roc_curve_utilities.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    selected = []
+    for r2 in roc_sorted.itertuples():
+        while len(selected) >= 2:
+            r1 = selected[-1]
+            r0 = selected[-2]
+            # Calculate the y value at r2's x if the slope between r0 and r1 is continued.
+            # If that y value is not larger than r2's actual y value we know that r1 lies
+            # below the line between r0 and r2 and can be dropped since we can reach all points
```
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/_roc_curve_utilities.py]: this is very different now, please see whether it makes more sense to you now

File: fairlearn/post_processing/_roc_curve_utilities.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            # between r0 and r2 through interpolation.
+            if (r1.y - r0.y) * (r2.x - r0.x) <= (r2.y - r0.y) * (r1.x - r0.x):
+                selected.pop()
```
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/_roc_curve_utilities.py]: good point

File: fairlearn/post_processing/_roc_curve_utilities.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    # labels provides better accuracy.
+    i = 0
+    count = [0, 0]
+    while i < n:
+        threshold = scores[i]
+        while scores[i] == threshold:
+            count[labels[i]] += 1
+            i += 1
+        # For the ROC curve we calculate points (x, y), where x represents
+        # the conditional probability P[Y_hat=0 | Y=0] and y represents
```
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/post_processing/_roc_curve_utilities.py]: changed! Not sure how I got to that...

File: setup.cfg [hunk 9]
```diff
@@ -0,0 +1,2 @@
+[flake8]
+max-line-length = 99
```
Comment (riedgar-ms) [inline setup.cfg]: Are we sure?

Comment (romanlutz) [reply setup.cfg]: My personal opinion is that restricting line lengths is stupid because line wrap exists, so why hard code when the line ends?
It's basically just a nuisance when writing code because you have to wrap and often make code harder to read which is the opposite of what flake8 wants you to achieve. This person expressed it pretty well, and in fact uses 119 like django (a huge project) https://simpleisbetterthancomplex.com/packages/2016/08/05/flake8.html

Comment (riedgar-ms) [reply setup.cfg]: Linewrap in Python is more dangerous than semicolon using languages :-)

I'm not averse to increasing the minimum (you'll see that some of my lines have `# noqa`), and I see that the Python SDK pushes it to 119. I'd keep to 99 for now.

**Comment by romanlutz** [comment]
Checking this in as a preliminary change. Still quite a few changes ahead, but this is a good checkpoint.
